core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala

```diff
@@ -482,40 +482,109 @@ class AutoTuner(
+        // Account for the setting of:
+        // appendRecommendation("spark.rapids.shuffle.multiThreaded.maxBytesInFlight", "4g")
+        executorMemOverhead += 4096L
+        setMaxBytesInFlight = true
+      }
+      // Pinned memory uses any unused space up to 4GB. Spill memory is same size as pinned.
+      val pinnedMem = Math.min(MAX_PINNED_MEMORY_MB,
+        (containerMemLeftOverOffHeap - executorMemOverhead) / 2).toLong
+      // Spill storage is set to the pinned size by default. Its not guaranteed to use just pinned
+      // memory though so the size worst case would be doesn't use any pinned memory and uses
+      // all off heap memory.
+      val spillMem = pinnedMem
+      if (containerMemLeftOverOffHeap >= executorMemOverhead + pinnedMem + spillMem) {
+        executorMemOverhead += pinnedMem + spillMem
+      } else {
+        // use min pinned and spill mem
+        executorMemOverhead += MIN_PINNED_MEMORY_MB + MIN_SPILL_MEMORY_MB
+      }
+      (pinnedMem, executorMemOverhead, executorHeap, setMaxBytesInFlight)
+    } else {
+      // otherwise we have to adjust heuristic of the executor heap size
+      // recommendedMinHeap = DEF_HEAP_PER_CORE_MB * numExecutorCores
+      // first calculate what we think min overhead is and make sure we have enough
+      // for that
+      // calculate minimum heap size
+      val minExecHeapMem = MIN_HEAP_PER_CORE_MB * numExecutorCores
+      if ((containerMem - minOverhead) < minExecHeapMem) {
+        // For now just throw so we don't get any tunings and its obvious to user this isn't a good
+        // setup. In the future we may just recommend them to use larger nodes. This would be more
+        // ideal once we hook up actual executor heap from an eventlog vs what user passes in.
+        throwNotEnoughMemException()
+        (0, 0, 0, false)
+      } else {
+        val leftOverMemUsingMinHeap = containerMem - minExecHeapMem
+        if (leftOverMemUsingMinHeap < 0) {
+          throwNotEnoughMemException()
+        }
+        // Pinned memory uses any unused space up to 4GB. Spill memory is same size as pinned.
+        val pinnedMem = Math.min(MAX_PINNED_MEMORY_MB, (leftOverMemUsingMinHeap / 2)).toLong
+        val spillMem = pinnedMem
+        // spill memory is by default same size as pinned memory
+        executorMemOverhead += pinnedMem + spillMem
+        (pinnedMem, executorMemOverhead, minExecHeapMem, setMaxBytesInFlight)
+      }
+    }
+  }
+
+  private def throwNotEnoughMemException(): Unit = {
+    // in the future it would be nice to enhance the error message with a recommendation of size
+    val msg = "This node/worker configuration is not ideal for using the Spark Rapids " +
+      "Accelerator because it doesn't have enough memory for the executors. " +
+      "We recommend using nodes/workers with more memory."
+    logError(msg)
+    throw new IllegalArgumentException(msg)
```

Comment:
Just double checking that this is the interntion:

- When AutoTuner throws the IllegalArgumentException, then no tuning output is generated at all
- Which try-catch is going to handle that exception? Does it cause the entire app to fail?

Reply:
The intention is to just fail the tuning but succeed the qualification.  This is being caught in TunerContext.tuneApplication.

```
24/06/06 08:44:11 ERROR TunerContext: Failed to generate tuning recommendations for app: app-20240603192440-0000
java.lang.IllegalArgumentException: This node/worker configuration is not ideal for using the Spark Rapids Accelerator because it doesn't h
ave enough memory for the executors. We recommend using nodes/workers with more memory.
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.throwNotEnoughMemException(AutoTuner.scala:587)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.calcOverallMemory(AutoTuner.scala:564)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.calculateClusterLevelRecommendations(AutoTuner.scala:716)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.getRecommendedProperties(AutoTuner.scala:1151)
        at com.nvidia.spark.rapids.tool.tuning.QualificationAutoTuner.runAutoTuner(QualificationAutoTuner.scala:71)
        at com.nvidia.spark.rapids.tool.tuning.TunerContext$$anonfun$tuneApplication$1.$anonfun$applyOrElse$1(TunerContext.scala:63)
        at scala.util.Try$.apply(Try.scala:213)
        at com.nvidia.spark.rapids.tool.tuning.TunerContext$$anonfun$tuneApplication$1.applyOrElse(TunerContext.scala:63)
        at com.nvidia.spark.rapids.tool.tuning.TunerContext$$anonfun$tuneApplication$1.applyOrElse(TunerContext.scala:60)
        at scala.PartialFunction$Lifted.apply(PartialFunction.scala:228)
        at scala.PartialFunction$Lifted.apply(PartialFunction.scala:224)
        at scala.Option.collect(Option.scala:432)
        at com.nvidia.spark.rapids.tool.tuning.TunerContext.tuneApplication(TunerContext.scala:60)
        at com.nvidia.spark.rapids.tool.qualification.Qualification.$anonfun$qualifyApp$3(Qualification.scala:177)
        at scala.Option.foreach(Option.scala:407)
        at com.nvidia.spark.rapids.tool.qualification.Qualification.com$nvidia$spark$rapids$tool$qualification$Qualification$$qualifyApp(Qu
alification.scala:172)
        at com.nvidia.spark.rapids.tool.qualification.Qualification$QualifyThread.run(Qualification.scala:60)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:750)

```

Reply:
profiler handles it as well:

```
java.lang.IllegalArgumentException: This node/worker configuration is not ideal for using the Spark Rapids Accelerator because it doesn't have enough memory for the executors. We recommend using nodes/workers with more memory.
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.throwNotEnoughMemException(AutoTuner.scala:587)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.calcOverallMemory(AutoTuner.scala:564)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.calculateClusterLevelRecommendations(AutoTuner.scala:716)
        at com.nvidia.spark.rapids.tool.profiling.AutoTuner.getRecommendedProperties(AutoTuner.scala:1151)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.runAutoTuner(Profiler.scala:423)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.$anonfun$writeOutput$55(Profiler.scala:551)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.$anonfun$writeOutput$55$adapted(Profiler.scala:498)
        at scala.collection.immutable.List.foreach(List.scala:431)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.writeOutput(Profiler.scala:498)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.com$nvidia$spark$rapids$tool$profiling$Profiler$$writeSafelyToOutput(Profiler.scala:567)
        at com.nvidia.spark.rapids.tool.profiling.Profiler$ProfileProcessThread$1.$anonfun$run$3(Profiler.scala:275)
        at com.nvidia.spark.rapids.tool.profiling.Profiler$ProfileProcessThread$1.$anonfun$run$3$adapted(Profiler.scala:270)
        at com.nvidia.spark.rapids.tool.profiling.Profiler.com$nvidia$spark$rapids$tool$profiling$Profiler$$profileApp(Profiler.scala:182)
        at com.nvidia.spark.rapids.tool.profiling.Profiler$ProfileProcessThread$1.run(Profiler.scala:270)

```